<a href="https://colab.research.google.com/github/heysourin/Machine-Learning-Projects/blob/main/Credit%20card%20fraud/Credit_card_fraud_machine_learning_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [9]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [3]:
df = pd.read_csv("creditcard.csv")

In [4]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62    0.0  
1  0.125895 -0.008983  0.014724    2.69    0.0  
2 -0.139097 -0.055353 -0.059752  378.66    0.0  
3 -0.221929  0.062723  0.061458  123.50    0.0  
4  0.502292  0.219422  0.215153   69.99    0.0  

[5 rows x 31 columns]

In [10]:
df.tail()

Time        V1        V2        V3        V4        V5        V6  \
257452  158172.0 -1.120369  0.501011 -1.529317  0.695992  2.181396 -0.997850   
257453  158174.0 -0.156508  0.656963 -0.135510 -0.120234 -0.110735  0.026058   
257454  158175.0  2.063160  0.057377 -1.226656  0.351724  0.050100 -1.223581   
257455  158176.0 -1.266069  1.182223 -0.639260  0.979701  0.210889 -0.251421   
257456  158176.0 -0.873452  0.602545 -0.160409 -0.477630  0.316238 -1.148471   

              V7        V8        V9  ...       V21       V22       V23  \
257452  1.100139 -0.126678 -0.323019  ... -0.141843  0.281271  0.571184   
257453  0.146752  0.530776  0.563762  ... -0.266351 -0.710929  0.521638   
257454  0.268653 -0.427801  0.462757  ...  0.278107  1.036505 -0.006535   
257455  0.337728  0.770269 -1.263478  ...  0.298574  0.515923 -0.531365   
257456  0.905051 -0.015166  0.081742  ... -0.275979 -0.565464  0.090000   

             V24       V25       V26       V27       V28  Amount  Class  
257452  0.138720  0.539882 -0.568312 -0.275066 -0.413842    1.00    0.0  
257453  0.568582 -1.221039 -0.072223 -0.000540  0.088254   72.90    0.0  
257454  0.056109  0.277374 -0.102485 -0.005481 -0.055708    1.00    0.0  
257455  0.736880  0.896669  0.447270 -0.330024 -0.142205   37.97    0.0  
257456       NaN       NaN       NaN       NaN       NaN     NaN    NaN  

[5 rows x 31 columns]



---
### **Note:**

- Amount is in USD
- Class 0 --> Legit transaction, 1 --> Fradualnt transaction
- The dataset contains 158176 seconds of data, means around 44 hours of data.
- It is a very unbalanced dataset.

---



In [5]:
df.shape

(257457, 31)

In [6]:
df.isna().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       1
V25       1
V26       1
V27       1
V28       1
Amount    1
Class     1
dtype: int64

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 257457 entries, 0 to 257456
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    257457 non-null  float64
 1   V1      257457 non-null  float64
 2   V2      257457 non-null  float64
 3   V3      257457 non-null  float64
 4   V4      257457 non-null  float64
 5   V5      257457 non-null  float64
 6   V6      257457 non-null  float64
 7   V7      257457 non-null  float64
 8   V8      257457 non-null  float64
 9   V9      257457 non-null  float64
 10  V10     257457 non-null  float64
 11  V11     257457 non-null  float64
 12  V12     257457 non-null  float64
 13  V13     257457 non-null  float64
 14  V14     257457 non-null  float64
 15  V15     257457 non-null  float64
 16  V16     257457 non-null  float64
 17  V17     257457 non-null  float64
 18  V18     257457 non-null  float64
 19  V19     257457 non-null  float64
 20  V20     257457 non-null  float64
 21  V21     25

In [12]:
df = df.dropna()

In [13]:
df.isna().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [15]:
# Chceking the distribution of legit and fradulant transaction.
df['Class'].value_counts() #Unbalanced dataset, model will not be able to recognize, so have to handle the unbalance dataset.

Class
0.0    256986
1.0       470
Name: count, dtype: int64

In [16]:
# Separating the data for analysis
legit = df[df['Class'] == 0]
fraud =  df[df['Class'] == 1]

In [17]:
legit.shape, fraud.shape

((256986, 31), (470, 31))

In [20]:
# Statistical measures for the data
legit['Amount'].describe()

count    256986.000000
mean         90.251211
std         250.331175
min           0.000000
25%           5.990000
50%          22.950000
75%          79.070000
max       19656.530000
Name: Amount, dtype: float64

In [22]:
fraud['Amount'].describe() # avg spent is more here.

count     470.000000
mean      122.636404
std       260.193766
min         0.000000
25%         1.000000
50%         8.770000
75%       104.615000
max      2125.870000
Name: Amount, dtype: float64

In [24]:
# compare the values of both tx
df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0.0    87403.372666 -0.016966 -0.024598  0.090264  0.009930 -0.024115   
1.0    76838.787234 -4.916086  3.753249 -7.167910  4.618699 -3.319744   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.016705 -0.002299  0.000402  0.005217  ...  0.002960 -0.002402   
1.0   -1.384064 -5.737824  0.615789 -2.626903  ...  0.368304  0.742969   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.007955 -0.004477  0.001341  0.016002  0.001056 -0.000933  0.000536   
1.0    0.002737 -0.052985 -0.105980  0.046062  0.048818  0.162808  0.075408   

           Amount  
Class              
0.0     90.251211  
1.0    122.636404  

[2 rows x 30 columns]

Under-Sampling
- Build a sample dataset containing sample similar distribution of legit and fraud Tx
- Total fraud tx = 492
- We will make a dataset of 492 legit + 492 fraud Tx

In [25]:
legit_sample = legit.sample(n=492)

In [26]:
# Merging
new_df = pd.concat([legit_sample, fraud], axis=0)

In [27]:
new_df.head()

Time        V1        V2        V3        V4        V5        V6  \
79268    57945.0  1.239009  0.048368  0.353644  0.446088 -0.435264 -0.524232   
7472     10162.0 -2.118628 -1.665831  2.168097  0.646847  1.995953 -0.208324   
1404      1084.0 -0.327426  0.377346  1.044564 -0.821649 -0.938044 -0.416148   
27283    34482.0 -2.055826 -0.994801  0.773803 -0.113505  2.180081 -1.692024   
194306  130525.0  1.721814 -1.720044 -1.596695 -0.667331 -0.526829  0.154072   

              V7        V8        V9  ...       V21       V22       V23  \
79268  -0.180653  0.073928  0.104446  ... -0.250004 -0.853593  0.106226   
7472   -0.340295 -0.557743  2.636188  ... -0.322477  0.517332 -0.869818   
1404    2.082513 -0.958191 -1.466671  ... -0.774607 -1.402097  0.037759   
27283   0.633078 -0.649401  0.563970  ... -0.296783  0.387778  0.425360   
194306 -0.388302 -0.119609 -0.193806  ...  0.629430  1.574427 -0.398673   

             V24       V25       V26       V27       V28  Amount  Class  
79268  -0.061137  0.165469  0.104310 -0.046140  0.001297    1.98    0.0  
7472   -0.748196 -0.481247 -0.518205 -0.103893 -0.627536  147.86    0.0  
1404    0.334190 -0.311227  0.054431 -0.448106 -0.465675  229.90    0.0  
27283  -0.032020 -0.086519  0.306585  0.221242 -0.101605    0.00    0.0  
194306  0.498809  0.451013  0.235393 -0.063326 -0.035086  250.00    0.0  

[5 rows x 31 columns]

In [29]:
new_df.tail()

Time        V1        V2        V3        V4        V5        V6  \
252774  155965.0 -1.201398  4.864535 -8.328823  7.652399 -0.167445 -2.767695   
254344  156685.0 -0.129778  0.141547 -0.894702 -0.457662  0.810608 -0.504723   
254395  156710.0  0.202402  1.176270  0.346379  2.882138  1.407133 -0.504355   
255403  157207.0  1.170756  2.501038 -4.986159  5.374160  0.997798 -1.259004   
255556  157284.0 -0.242245  4.147186 -5.672349  6.493741  1.591168 -1.602523   

              V7        V8        V9  ...       V21       V22       V23  \
252774 -3.176421  1.623279 -4.367228  ...  0.532320 -0.556913  0.192444   
254344  1.373588 -0.209476  0.208494  ... -0.032643 -0.246526  0.484108   
254395  1.438537 -0.395603 -1.555142  ...  0.242560  0.841230 -0.370157   
255403 -1.237689  0.358426 -2.612489  ...  0.123145 -0.713201 -0.080868   
255556 -0.950463  0.722903 -4.128505  ...  0.249023 -0.480286 -0.286080   

             V24       V25       V26       V27       V28  Amount  Class  
252774 -0.698588  0.025003  0.514968  0.378105 -0.053133    0.77    1.0  
254344  0.359637 -0.435972 -0.248480  0.021527  0.109192  187.11    1.0  
254395 -0.026012  0.491954  0.234576 -0.279788 -0.331933    7.59    1.0  
255403 -0.964310  0.338568  0.068630  0.481588  0.268226    4.97    1.0  
255556 -1.153575 -0.035571  0.559628  0.409446  0.221048    0.77    1.0  

[5 rows x 31 columns]

In [28]:
new_df.shape

(962, 31)

In [30]:
new_df['Class'].value_counts() # see we have 492 fraud and 492 legit tx

Class
0.0    492
1.0    470
Name: count, dtype: int64

In [32]:
new_df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0.0    86423.603659 -0.131208 -0.016308  0.161322  0.066359 -0.077777   
1.0    76838.787234 -4.916086  3.753249 -7.167910  4.618699 -3.319744   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.118189  0.061628 -0.131770 -0.027106  ...  0.013602 -0.029529   
1.0   -1.384064 -5.737824  0.615789 -2.626903  ...  0.368304  0.742969   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.025306  0.026325 -0.027264 -0.008152  0.004649 -0.038439 -0.026473   
1.0    0.002737 -0.052985 -0.105980  0.046062  0.048818  0.162808  0.075408   

           Amount  
Class              
0.0    109.208110  
1.0    122.636404  

[2 rows x 30 columns]

In [33]:
# Splitting X and y
X = new_df.drop(columns='Class', axis=1)
y = new_df['Class']

In [34]:
X

Time        V1        V2        V3        V4        V5        V6  \
79268    57945.0  1.239009  0.048368  0.353644  0.446088 -0.435264 -0.524232   
7472     10162.0 -2.118628 -1.665831  2.168097  0.646847  1.995953 -0.208324   
1404      1084.0 -0.327426  0.377346  1.044564 -0.821649 -0.938044 -0.416148   
27283    34482.0 -2.055826 -0.994801  0.773803 -0.113505  2.180081 -1.692024   
194306  130525.0  1.721814 -1.720044 -1.596695 -0.667331 -0.526829  0.154072   
...          ...       ...       ...       ...       ...       ...       ...   
252774  155965.0 -1.201398  4.864535 -8.328823  7.652399 -0.167445 -2.767695   
254344  156685.0 -0.129778  0.141547 -0.894702 -0.457662  0.810608 -0.504723   
254395  156710.0  0.202402  1.176270  0.346379  2.882138  1.407133 -0.504355   
255403  157207.0  1.170756  2.501038 -4.986159  5.374160  0.997798 -1.259004   
255556  157284.0 -0.242245  4.147186 -5.672349  6.493741  1.591168 -1.602523   

              V7        V8        V9  ...       V20       V21       V22  \
79268  -0.180653  0.073928  0.104446  ... -0.196405 -0.250004 -0.853593   
7472   -0.340295 -0.557743  2.636188  ... -0.832133 -0.322477  0.517332   
1404    2.082513 -0.958191 -1.466671  ... -0.450602 -0.774607 -1.402097   
27283   0.633078 -0.649401  0.563970  ... -0.210249 -0.296783  0.387778   
194306 -0.388302 -0.119609 -0.193806  ...  0.511270  0.629430  1.574427   
...          ...       ...       ...  ...       ...       ...       ...   
252774 -3.176421  1.623279 -4.367228  ...  0.597026  0.532320 -0.556913   
254344  1.373588 -0.209476  0.208494  ...  0.399447 -0.032643 -0.246526   
254395  1.438537 -0.395603 -1.555142  ... -0.206239  0.242560  0.841230   
255403 -1.237689  0.358426 -2.612489  ...  0.227618  0.123145 -0.713201   
255556 -0.950463  0.722903 -4.128505  ...  0.562030  0.249023 -0.480286   

             V23       V24       V25       V26       V27       V28  Amount  
79268   0.106226 -0.061137  0.165469  0.104310 -0.046140  0.001297    1.98  
7472   -0.869818 -0.748196 -0.481247 -0.518205 -0.103893 -0.627536  147.86  
1404    0.037759  0.334190 -0.311227  0.054431 -0.448106 -0.465675  229.90  
27283   0.425360 -0.032020 -0.086519  0.306585  0.221242 -0.101605    0.00  
194306 -0.398673  0.498809  0.451013  0.235393 -0.063326 -0.035086  250.00  
...          ...       ...       ...       ...       ...       ...     ...  
252774  0.192444 -0.698588  0.025003  0.514968  0.378105 -0.053133    0.77  
254344  0.484108  0.359637 -0.435972 -0.248480  0.021527  0.109192  187.11  
254395 -0.370157 -0.026012  0.491954  0.234576 -0.279788 -0.331933    7.59  
255403 -0.080868 -0.964310  0.338568  0.068630  0.481588  0.268226    4.97  
255556 -0.286080 -1.153575 -0.035571  0.559628  0.409446  0.221048    0.77  

[962 rows x 30 columns]

In [35]:
y

79268     0.0
7472      0.0
1404      0.0
27283     0.0
194306    0.0
         ... 
252774    1.0
254344    1.0
254395    1.0
255403    1.0
255556    1.0
Name: Class, Length: 962, dtype: float64

In [37]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=2)

In [38]:
X.shape, X_train.shape, X_test.shape

((962, 30), (769, 30), (193, 30))

## **Model Training**

In [39]:
model = LogisticRegression()

In [40]:
model.fit(X_train, y_train)

LogisticRegression()

### Model evaluation:

In [46]:
# Accuracy on training data
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(y_train, X_train_prediction)
training_data_accuracy

0.94148244473342

In [45]:
# Test on training data
X_test_prediction = model.predict(X_test)
training_data_test = accuracy_score(y_test, X_test_prediction)
training_data_test

0.9378238341968912